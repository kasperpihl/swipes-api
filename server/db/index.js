import config from 'config';
import r from 'rethinkdb';
import Promise from 'bluebird';

const dbConfig = config.get('database');
const isCursor = (obj) => {
  // TODO: Suggest an r.isCursor() API or something similar.
  return obj != null && obj._conn != null;
};
const handleCursors = (results) => {
  return new Promise((resolve, reject) => {
    if (isCursor(results)) {
      return results.toArray()
        .then((array) => {
          // Clean up memory from the cursor
          results.close();

          return resolve(array);
        }).error((err) => {
          results.close();

          return reject(err);
        });
    }

    return resolve(results);
  });
};
const rethinkdb = {
  // options:
  // cursor: true/false(default) Most of the time we will
  // need just the results as array from the cursor
  // feed: true/false (to handle change feeds - don't close the connection)
  // returnConnection: return the cursor and the connection in an array
  rethinkQuery: (query, options = {}) => {
    return new Promise((resolve, reject) => {
      let conn;

      r.connect(dbConfig).then((localConn) => {
        conn = localConn;

        return query.run(conn);
      }).then((cursor) => {
        if (!options.feed) {
          conn.close();
        }

        if (options.cursor || options.feed) {
          if (options.feed) {
            if (options.socket) {
              options.socket.on('close', () => {
                conn.close({ noreplyWait: false });
              });
            }

            if (options.returnConnection) {
              return resolve([cursor, conn]);
            }

            return resolve(cursor);
          }

          return resolve(cursor);
        }

        return handleCursors(cursor)
          .then((results) => {
            return resolve(results);
          }).error((err) => {
            return reject(err);
          });
      }).error((err) => {
        return reject(err);
      });
    });
  },
};

module.exports = rethinkdb;
