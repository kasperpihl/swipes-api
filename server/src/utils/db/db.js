import { Pool } from 'pg';

const pool = new Pool({
  host: 'workspace.cuaunhcnlbha.us-east-1.rds.amazonaws.com',
  port: 5432,
  user: 'postgres',
  password: 'GAze0UGQyj',
  database: 'workspace'
});

export const query = (text, params) => {
  if (typeof text === 'object' && text.text && text.values) {
    return pool.query(text.text, text.values);
  }
  return pool.query(text, params);
};

export const transaction = async queries => {
  if (!Array.isArray(queries)) {
    throw 'db.transaction expects an array of queries';
  }
  // note: we don't try/catch this because if connecting throws an exception
  // we don't need to dispose of the client (it will be undefined)
  const client = await pool.connect();
  const results = [];
  try {
    await client.query('BEGIN');

    for (let i = 0; i < queries.length; i++) {
      let query = queries[i];
      if (typeof query === 'function') {
        query = query(results);
      }

      let text = query;
      let values;
      if (typeof text === 'object') {
        text = query.text;
        values = query.values;
      }

      const res = await client.query(text, values);

      if (typeof query.onSuccess === 'function') {
        query.onSuccess(res);
      }
      results.push(res);
    }
    await client.query('COMMIT');
  } catch (e) {
    await client.query('ROLLBACK');
    throw e;
  } finally {
    client.release();
  }
  return results;
};
