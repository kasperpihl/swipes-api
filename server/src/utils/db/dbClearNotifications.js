import r from 'rethinkdb';
import dbRunQuery from 'src/utils/db/dbRunQuery';
import dbSendEvents from 'src/utils/db/dbSendEvents';

export default async (notificationIds) => {
  if(!Array.isArray(notificationIds)) {
    notificationIds = [notificationIds];
  }

  const q = r.table('notifications').getAll(...notificationIds).delete({
    return_changes: true,
  });

  const res = await dbRunQuery(q);

  dbSendEvents(res.changes.map(({ old_val }) => ({
    user_ids: [ old_val.user_id ],
    type: 'notification_deleted',
    data: {
      notification_id: old_val.id,
    }
  })))
}