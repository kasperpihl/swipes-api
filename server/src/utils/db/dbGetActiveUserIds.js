import r from 'rethinkdb';
import dbRunQuery from './dbRunQuery';

export default (orgId) => dbRunQuery(
  r.table('organizations').get(orgId)('active_users')
)