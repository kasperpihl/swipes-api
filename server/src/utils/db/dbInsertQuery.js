import r from 'rethinkdb';

export default (table, obj, options = {}) => {
  if(typeof options.return_changes === 'undefined') {
    options.return_changes = true;
  }
  if(!Array.isArray(obj))Â {
    obj = [obj];
  }
  obj.forEach((o) => {
    o.updated_at = r.now();
    o.created_at = r.now();
  });
  
  return r.table(table).insert(obj, options);
}