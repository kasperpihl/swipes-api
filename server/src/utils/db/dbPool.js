const { Pool } = require('pg');

const pool = new Pool({
  host: 'workspace.cuaunhcnlbha.us-east-1.rds.amazonaws.com',
  port: 5432,
  user: 'postgres',
  password: 'GAze0UGQyj',
  database: 'workspace',
});

module.exports = {
  query: (text, params) => pool.query(text, params),
};
