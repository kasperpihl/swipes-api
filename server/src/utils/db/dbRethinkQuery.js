import r from 'rethinkdb';

const dbConfig = {
  host: 'rethinkdb-live5931.cloudapp.net',
  port: 28015,
  db: 'swipes',
  user: 'swipes-live',
  password: 'st@yProduct1v3Swipesl1v3'
};

const isCursor = obj => {
  // TODO: Suggest an r.isCursor() API or something similar.
  return obj != null && obj._conn != null;
};
const handleCursors = (results, conn) => {
  return new Promise((resolve, reject) => {
    if (isCursor(results)) {
      return results
        .toArray()
        .then(array => {
          // Clean up memory from the cursor
          results.close();

          return conn.close().then(() => {
            return resolve(array);
          });
        })
        .error(err => {
          results.close();

          return conn.close().then(() => {
            return reject(err);
          });
        });
    }

    return conn.close().then(() => {
      return resolve(results);
    });
  });
};
// options:
// cursor: true/false(default) Most of the time we will
// need just the results as array from the cursor
// feed: true/false (to handle change feeds - don't close the connection)
// returnConnection: return the cursor and the connection in an array
// dbConfig: pass a config object for connection to the db
export default (query, options = {}) =>
  new Promise((resolve, reject) => {
    let conn = null;

    return r
      .connect(dbConfig)
      .then(localConn => {
        conn = localConn;

        return query.run(conn);
      })
      .then(cursor => {
        if (options.cursor) {
          return conn.close().then(() => {
            return resolve(cursor);
          });
        }

        if (options.feed) {
          if (options.socket) {
            options.socket.on('close', () => {
              conn.close({ noreplyWait: false });
            });
          }

          if (options.returnConnection) {
            return resolve([cursor, conn]);
          }

          return resolve(cursor, conn);
        }

        return handleCursors(cursor, conn)
          .then(results => {
            if (results.errors > 0) {
              // From rethinkdb docs
              // first_error: If errors were encountered, contains the text of the first error
              return reject(results.first_error);
            }

            return resolve(results);
          })
          .catch(err => {
            return reject(err);
          });
      })
      .catch(err => {
        return reject(err);
      });
  });
