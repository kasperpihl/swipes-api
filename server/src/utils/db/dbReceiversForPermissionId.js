import { query } from './db';
export default async function dbReceiversForPermissionId(permissionId) {
  const receiverRes = await query(
    `
      SELECT granted_to
      FROM permissions
      WHERE permission_from = $1
    `,
    [permissionId]
  );
  if (!receiverRes.rows.length) {
    throw Error('Not found').code(404);
  }
  return receiverRes.rows.map(row => row.granted_to);
}
