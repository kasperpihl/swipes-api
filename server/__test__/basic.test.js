import r from 'rethinkdb';
import Promise from 'bluebird';
import rp from 'request-promise';
import db from '../db/';

const rpap = rp.defaults({
  // baseUrl: 'https://staging.swipesapp.com/v1',
  baseUrl: 'http://localhost:5000/v1',
  json: true,
});
const locals = {};

// const deleteProps
const updateLocals = (res, fields) => {
  fields.forEach((field) => {
    if (typeof field === 'object') {
      locals[field.key] = res[field.value];
      delete res[field.value];
    } else {
      locals[field] = res[field];
      delete res[field];
    }
  });

  return res;
};

const clearDatabase = () => {
  const organization_id = locals.organization.id;
  const clearUserQ = r.table('users').filter({
    email: 'tihomir+jest@swipesapp.com',
  }).delete();
  const clearUser2Q = r.table('users').filter({
    email: 'tihomir+jest2@swipesapp.com',
  }).delete();
  const clearUser3Q = r.table('users').filter({
    email: 'tihomir+jest3@swipesapp.com',
  }).delete();
  const clearOrganizationQ = r.table('organizations').get(organization_id).delete();
  const clearPostsQ = r.table('posts').filter({ organization_id }).delete();
  const clearMilestonesQ = r.table('milestones').filter({ organization_id }).delete();
  const clearGoalsQ = r.table('goals').filter({ organization_id }).delete();
  const promises = [
    db.rethinkQuery(clearUserQ),
    db.rethinkQuery(clearUser2Q),
    db.rethinkQuery(clearUser3Q),
    db.rethinkQuery(clearOrganizationQ),
    db.rethinkQuery(clearPostsQ),
    db.rethinkQuery(clearMilestonesQ),
    db.rethinkQuery(clearGoalsQ),
  ];

  console.log('CLEANING');

  return Promise.all(promises);
};

afterAll(() => {
  return clearDatabase();
});

test('sign up new user', () => {
  return rpap.post({
    url: '/users.signup',
    body: {
      email: 'tihomir+jest@swipesapp.com',
      password: 'jestit',
      first_name: 'Jest',
      last_name: 'Jestov',
    },
  })
    .then((result) => {
      expect(result).toMatchObject({
        token: expect.any(String),
        user_id: expect.any(String),
      });
      expect(updateLocals(result, ['token', 'user_id'])).toMatchSnapshot();
    });
});

test('creating organization', () => {
  return rpap.post({
    url: '/organizations.create',
    body: {
      organization_name: 'Jest Org',
      token: locals.token,
    },
  })
    .then((result) => {
      expect(result).toMatchObject({
        organization: expect.any(Object),
      });
      expect(updateLocals(result, ['organization'])).toMatchSnapshot();
    });
}, 60000);

test('inviting user', () => {
  return rpap.post({
    url: '/organizations.inviteUser',
    body: {
      organization_id: locals.organization.id,
      first_name: 'Jest2',
      email: 'tihomir+jest2@swipesapp.com',
      token: locals.token,
    },
  })
    .then((result) => {
      expect(result).toMatchObject({
        user: expect.any(Object),
        invitation_token: expect.any(String),
        organization: expect.any(Object),
      });
      expect(updateLocals(result, ['user', 'invitation_token', 'organization'])).toMatchSnapshot();
    });
}, 10000);

test('sign up with invitation token', () => {
  return rpap.post({
    url: '/users.signup',
    body: {
      email: 'tihomir+jest2@swipesapp.com',
      password: 'jestit2',
      first_name: 'Jest2',
      last_name: 'Jestov2',
      invitation_token: locals.invitation_token,
    },
  })
    .then((result) => {
      expect(result).toMatchObject({
        token: expect.any(String),
        user_id: expect.any(String),
      });

      delete result.user_id;
      delete result.token;

      expect(result).toMatchSnapshot();
    });
});

test('organization join', () => {
  let user_3_token = null;

  return rpap.post({
    url: '/users.signup',
    body: {
      email: 'tihomir+jest3@swipesapp.com',
      password: 'jestit',
      first_name: 'Jest',
      last_name: 'Jestov',
    },
  })
    .then((result) => {
      user_3_token = result.token;

      return rpap.post({
        url: '/organizations.inviteUser',
        body: {
          organization_id: locals.organization.id,
          first_name: 'Jest2',
          email: 'tihomir+jest3@swipesapp.com',
          token: locals.token,
        },
      });
    })
    .then((result) => {
      return rpap.post({
        url: '/organizations.join',
        body: {
          organization_id: locals.organization.id,
          token: user_3_token,
        },
      })
        .then((result) => {
          expect(result).toMatchObject({
            organization: expect.any(Object),
          });
          expect(updateLocals(result, ['organization'])).toMatchSnapshot();
        });
    });
}, 20000);

test('init', () => {
  return rpap.post({
    url: '/init',
    body: {
      timestamp: null,
      token: locals.token,
    },
  })
    .then((result) => {
      expect(result).toMatchObject({
        me: expect.any(Object),
        timestamp: expect.any(String),
        full_fetch: expect.any(Boolean),
        sofi: expect.any(Object),
        // users: expect.any(Array),
        // goals: expect.any(Array),
        // milestones: expect.any(Array),
        // ways: expect.any(Array),
        // notes: expect.any(Array),
        // posts: expect.any(Array),
        // services: expect.any(Array),
        // notifications: expect.any(Array),
        // onboarding: expect.any(Array),
        // pending_organizations: expect.any(Array),
      });
    });
}, 10000);

test('milestone create', () => {
  return rpap.post({
    url: '/milestones.create',
    body: {
      title: 'Test milestone',
      organization_id: locals.organization.id,
      token: locals.token,
    },
  })
    .then((result) => {
      const milestone = updateLocals(result.milestone, [{
        key: 'milestone_id',
        value: 'id',
      }, 'updated_at', 'created_at', 'created_by', 'organization_id', 'history']);

      result.milestone = milestone;
      expect(result).toMatchSnapshot();
    });
});

test('goal create', () => {
  return rpap.post({
    url: '/goals.create',
    body: {
      goal: {
        title: 'Test goal',
        assignees: [locals.user_id],
      },
      organization_id: locals.organization.id,
      token: locals.token,
    },
  })
    .then((result) => {
      const goal = updateLocals(result.goal, [{
        key: 'goal_id',
        value: 'id',
      }, 'updated_at', 'created_at', 'created_by', 'organization_id', 'history', 'assignees']);

      result.goal = goal;
      expect(result).toMatchSnapshot();
    });
}, 10000);

test('post create type post', () => {
  return rpap.post({
    url: '/posts.create',
    body: {
      message: 'test post',
      organization_id: locals.organization.id,
      type: 'post',
      token: locals.token,
    },
  })
    .then((result) => {
      const post = updateLocals(result.post, [{
        key: 'post_id',
        value: 'id',
      }, 'updated_at', 'created_at', 'created_by', 'organization_id', 'followers']);

      result.post = post;
      expect(result).toMatchSnapshot();
    });
}, 10000);

test('sign in', () => {
  return rpap.post({
    url: '/users.signin',
    body: {
      email: 'tihomir+jest@swipesapp.com',
      password: 'jestit',
    },
  })
    .then((result) => {
      expect(result).toMatchObject({
        token: expect.any(String),
      });

      expect(updateLocals(result, ['token'])).toMatchSnapshot();
    });
});

test('sign out', () => {
  return rpap.post({
    url: '/users.signout',
    body: {
      token: locals.token,
    },
  })
    .then((result) => {
      expect(result).toMatchSnapshot();
    });
});
