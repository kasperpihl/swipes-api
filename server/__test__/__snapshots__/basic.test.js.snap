// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creating organization 1`] = `
Object {
  "ok": true,
}
`;

exports[`goal create 1`] = `
Object {
  "goal": Object {
    "archived": false,
    "attachment_order": Array [],
    "attachments": Object {},
    "completed_at": null,
    "milestone_id": null,
    "step_order": Array [],
    "steps": Object {},
    "title": "Test goal",
  },
  "ok": true,
}
`;

exports[`inviting user 1`] = `
Object {
  "ok": true,
}
`;

exports[`milestone create 1`] = `
Object {
  "milestone": Object {
    "closed_at": null,
    "due_date": null,
    "goal_order": Array [],
    "title": "Test milestone",
  },
  "ok": true,
}
`;

exports[`post create type post 1`] = `
Object {
  "ok": true,
  "post": Object {
    "archived": false,
    "attachments": Array [],
    "comments": Object {},
    "context": Object {},
    "message": "test post",
    "reactions": Array [],
    "tagged_users": Array [],
    "type": "post",
  },
}
`;

exports[`sign in 1`] = `
Object {
  "ok": true,
}
`;

exports[`sign out 1`] = `
Object {
  "ok": true,
}
`;

exports[`sign up new user 1`] = `
Object {
  "ok": true,
}
`;

exports[`sign up with invitation token 1`] = `
Object {
  "ok": true,
}
`;
